pgrx-sql-entity-graph/src/control_file.rs

```diff
@@ -43,18 +45,46 @@ pub struct ControlFile {
 }
 
 impl ControlFile {
-    /// Parse a `.control` file.
+    /// Parse a `.control` file, performing all known pgrx dynamic variable substitutions.
+    ///
+    /// # Supported Dynamic Variable Substitutions
+    ///
+    /// `@CARGO_VERSION@`:  Replaced with the value of the environment variable `CARGO_PKG_VERSION`,
+    ///                     which is set by cargo, or failing that, `cargo-pgrx` using the package
+    ///                     version from the extension's `Cargo.toml` file
+    ///
+    /// # Errors
+    ///
+    /// Returns a `ControlFileError` if any of the required fields are missing from the input string
+    /// or if any required environment variables (for dynamic variable substitution) are missing
     ///
     /// ```rust
     /// use pgrx_sql_entity_graph::ControlFile;
     /// # fn main() -> eyre::Result<()> {
-    /// let context = include_str!("../../pgrx-examples/custom_types/custom_types.control");
+    /// # // arrays.control chosen because it does **NOT** use the @CARGO_VERSION@ variable
+    /// let context = include_str!("../../pgrx-examples/arrays/arrays.control");
     /// let _control_file = ControlFile::from_str(context)?;
     /// # Ok(())
     /// # }
     /// ```
     #[allow(clippy::should_implement_trait)]
     pub fn from_str(input: &str) -> Result<Self, ControlFileError> {
+        fn do_var_replacements(mut input: String) -> Result<String, ControlFileError> {
+            const CARGO_VERSION: &str = "@CARGO_VERSION@";
+
+            // endeavor to not require external values if they're not used by the input
+            if input.contains(CARGO_VERSION) {
+                input = input.replace(
+                    CARGO_VERSION,
+                    &std::env::var("CARGO_PKG_VERSION").map_err(|_| {
+                        ControlFileError::MissingEnvvar("CARGO_PKG_VERSION".to_string())
+                    })?,
+                );
+            }
```

Comment:
This does introduce a slight behavior divergence, but that seems fine, since the main validation effort is done via cargo-pgrx.